pallets/ds-maps/src/lib.rs

```diff
@@ -308,9 +384,441 @@ impl Area {
     } 
 }
 
+#[derive(Encode, Decode, Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Page<Coord> {
+    pub bitmap: [[RootId; PAGE_WIDTH as usize]; PAGE_LENGTH as usize],
+    _phantom: PhantomData<Coord>,
+}
+
+impl<
+    Coord: Default
+    + FromStr
+    + Copy
+    + CastToType
+> Page<Coord> {
+    fn new() -> Self {
+        Page {
+            bitmap: [[0u64; PAGE_WIDTH as usize]; PAGE_LENGTH as usize],
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Calculates the number of pages to extract from the storage using the coordinates
+    pub fn get_amount_of_pages_to_extract_using_box(bounding_box: Box3D<Coord>) -> u32 {
+        let (sw_row_index, sw_column_index) = Self::get_cell_indexes(bounding_box.south_west);
+        let (ne_row_index, ne_column_index) = Self::get_cell_indexes(bounding_box.north_east);
+
+        Self::get_amount_of_pages_to_extract(sw_row_index, sw_column_index, ne_row_index, ne_column_index)
+    }
+
+    /// Calculates the number of pages to extract from the storage using the indexes
+    pub fn get_amount_of_pages_to_extract(
+        sw_row_index: u32, sw_column_index: u32,
+        ne_row_index: u32, ne_column_index: u32,
+    ) -> u32 {
+        let sw_cell_page_index = Self::get_index(sw_row_index, sw_column_index);
+        let ne_cell_page_index = Self::get_index(ne_row_index, ne_column_index);
+
+        if sw_cell_page_index == ne_cell_page_index {
+            let amount_of_pages: u32 = 1;
+            return amount_of_pages;
+        }
+
+        let (sw_page_row_index, sw_page_column_index) = Self::extract_values_from_page_index(sw_cell_page_index);
+        let (ne_page_row_index, ne_page_column_index) = Self::extract_values_from_page_index(ne_cell_page_index);
+
+        if ne_page_row_index < sw_page_row_index || ne_page_column_index < sw_page_column_index {
+            return 0;
+        }
+
+        let mut offset: u32 = 1;
+
+        if ne_page_row_index == sw_page_row_index {
+            if sw_column_index / PAGE_WIDTH == 0 {
+                return ne_page_column_index / PAGE_WIDTH;
+            }
+            return ne_page_column_index / PAGE_WIDTH - sw_page_column_index / PAGE_WIDTH + offset;
+        }
+
+        if sw_page_column_index == ne_page_column_index {
+            if sw_row_index / PAGE_LENGTH == 0 {
+                return ne_page_row_index / PAGE_LENGTH;
+            }
+            return ne_page_row_index / PAGE_LENGTH - sw_page_row_index / PAGE_LENGTH + offset;
+        }
+
+        if sw_row_index / PAGE_LENGTH == 0 && sw_column_index / PAGE_WIDTH == 0 {
+            offset = 2;
+        }
+
+        (ne_page_row_index / PAGE_LENGTH - sw_page_row_index / PAGE_LENGTH) +
+            (ne_page_column_index / PAGE_WIDTH - sw_page_column_index / PAGE_WIDTH)
+            + offset
+    }
+
+    /// Gets the indexes of the pages to be extracted
+    pub fn get_pages_indexes_to_be_extracted(
+        amount_of_pages_to_extract: u32,
```

Comment:
Isn't it better to call `get_amount_of_pages_to_extract` inside this function, as it actually depends on next 4 arguments?

Reply:
In this case, the function will calculate the value that was already found in the previous steps